quinn/src/connection.rs

```diff
@@ -859,37 +861,63 @@ where
 
     fn drive_timer(&mut self, cx: &mut Context) -> bool {
         let mut keep_going = false;
-        loop {
-            if let Some(ref mut delay) = self.timer {
-                if delay.poll_unpin(cx) == Poll::Ready(()) {
-                    // We must get a fresh `now` each iteration. If the value were cached and tokio
-                    // deems a timer for a later time than the cached value to be expired, we'd get
-                    // stuck in an infinite loop resetting it.
-                    self.inner.handle_timeout(Instant::now());
-                    self.timer = None;
-                    // A timer expired, so the caller needs to check for new transmits, which might
-                    // cause new timers to be set.
-                    keep_going = true;
+
+        // Check whether we need to (re)set the timer. If so, we must poll again to ensure the
+        // timer is registered with the runtime (and check whether it's already
+        // expired). Otherwise, exit the loop.
+        match self.inner.poll_timeout().map(TokioInstant::from_std) {
```

Comment:
Btw: I wasn't sure why the current code checked expirations before checking whether timers are still active.
The other way around seems more logical to me, since it also allows to avoid polling the timer.

Reply:
IIRC the motivation was that invoking `handle_timeout` may cause new timers to be set, although the outer `keep_going` loop probably makes it moot semantically.

Reply:
yes, that's handled by the outer loop